### ISSUE TITLE
UnityEditor crash 

### ISSUE THREAD

**Issue by Rurouni** [Cysharp/MagicOnion#27]
I am seeing seemingly random editor crash with such call stack:
```
0x00007FFF5879F7DB (ntdll) RtlpUnWaitCriticalSection
0x00007FFF587A0CE4 (ntdll) RtlEnterCriticalSection
0x00007FFF587A0C10 (ntdll) RtlEnterCriticalSection
0x00007FFF054A9BBD (grpc_csharp_ext) grpcsharp_test_nop
0x00007FFF0549568D (grpc_csharp_ext) grpcsharp_test_nop
0x00007FFF05491B35 (grpc_csharp_ext) grpcsharp_completion_queue_next
0x0000000036C73B3D (Mono JIT Code) (wrapper managed-to-native) Grpc.Core.Internal.NativeMethods/NativeCalls:grpcsharp_completion_queue_next (Grpc.Core.Internal.CompletionQueueSafeHandle)
0x0000000036C74A0D (Mono JIT Code) (wrapper delegate-invoke) Grpc.Core.Internal.NativeMethods/Delegates/grpcsharp_completion_queue_next_delegate:invoke_CompletionQueueEvent__this___CompletionQueueSafeHandle (Grpc.Core.Internal.CompletionQueueSafeHandle)
0x0000000036C72EC5 (Mono JIT Code) [C:\dev\SeedClient\Assets\Catapult\Network\gRPC\Core\Internal\CompletionQueueSafeHandle.cs:78] Grpc.Core.Internal.CompletionQueueSafeHandle:Next () 
0x0000000036C717B9 (Mono JIT Code) [C:\dev\SeedClient\Assets\Catapult\Network\gRPC\Core\Internal\GrpcThreadPool.cs:195] Grpc.Core.Internal.GrpcThreadPool:RunHandlerLoop (Grpc.Core.Internal.CompletionQueueSafeHandle,Grpc.Core.Profiling.IProfiler) 
0x0000000036C715F6 (Mono JIT Code) [C:\dev\SeedClient\Assets\Catapult\Network\gRPC\Core\Internal\GrpcThreadPool.cs:173] Grpc.Core.Internal.GrpcThreadPool/<CreateAndStartThread>c__AnonStorey0:<>m__0 () 
0x0000000001F07392 (Mono JIT Code) (wrapper runtime-invoke) object:runtime_invoke_void__this__ (object,intptr,intptr,intptr)
0x00007FFF0B345703 (mono) [c:\buildslave\mono\build\mono\mini\mini.c:4937] mono_jit_runtime_invoke 
0x00007FFF0B298425 (mono) [c:\buildslave\mono\build\mono\metadata\object.c:2623] mono_runtime_invoke 
0x00007FFF0B2C36E7 (mono) [c:\buildslave\mono\build\mono\metadata\threads.c:721] start_wrapper 
0x00007FFF0B3782D5 (mono) [c:\buildslave\mono\build\libgc\win32_threads.c:599] thread_start 
0x00007FFF55F38364 (KERNEL32) BaseThreadInitThunk
0x00007FFF587D70D1 (ntdll) RtlUserThreadStart
```
It seems to be connected to recreating/disposing channels as I reproduce it when running such test code in a loop:

``` CSharp
public IEnumerator TestPingPong()
        {
            for (int i = 0; i < 30; i++)
            {
                Log.Debug("Creating channel");
                Channel channel = UnitTestClient.CreateChannel();
                Log.Debug("Created channel");
                var protocolSerialiser = new SeedProtocolSerializer();
                var marshaller = Marshallers.Create<Message>(protocolSerialiser.ToBytes, protocolSerialiser.ToMessage);
                var method = new Method<Message, Message>(MethodType.DuplexStreaming,
                    "x", "x", marshaller, marshaller);

                var host = "localhost";
                var caller = new EditorWindowSupportsCallInvoker(channel);
                //var caller = new DefaultCallInvoker(channel);
                using (var call = caller.AsyncDuplexStreamingCall(method, host,
                    new CallOptions(deadline: DateTime.UtcNow.AddSeconds(5))))
                {
                    string data = "hello unity";
                    var request = new Ping() {Data = data}.AsMessage();
                    Log.Debug("calling service request:{$request}", request);
                    yield return call.RequestStream.WriteAsync(new Ping() {Data = data}.AsMessage())
                        .ToYieldInstruction();

                    var resultPromise = call.ResponseStream.AsObservable().Take(1).ToYieldInstruction();
                    yield return resultPromise;

                    resultPromise.HasResult.IsTrue();
                    var pong = resultPromise.Result as Message<Pong>;
                    Log.Debug("recieved reply:{reply}", pong);
                    pong.IsNotNull();
                    pong.Body.Data.Is(data + " back");
                    yield return call.RequestStream.CompleteAsync().ToYieldInstruction();
                }

                Log.Debug("Closing channel");
                try
                {
                    yield return channel.ShutdownAsync().ToYieldInstruction();
                }
                finally
                {
                    Log.Debug("Closed channel");
                }
            }

        }
```
It's definitely something to do with `channel.ShutdownAsync()` because if you comment it out there is no crash
I was wondering if it's something you have seen before

**Comment by neuecc** [comment]
Which line is it?
I know the problem of freezing when have a subscription.
Therefore, Thread.Join is suppressed.
https://github.com/grpc/grpc/issues/9202

If the subscription still remains, UnityEditor can not start again.
MagicOnionWindow is a window for check unsubscribe.

This is an example of secure cancellation support.

```csharp
var channel = new Channel("localhost", ChannelCredentials.Insecure);
var subscription = MagicOnionClient.Create<IFoo>(channel).BarServerStreaming().Subscribe();

// safety for UnityEditor stop
channel.RegisterStreamingSubscription(subscription);
```

**Comment by Rurouni** [comment]
Is there any way I can see the line? I have copied it the editor internal log file relevant to the crash. Unity doesn't hang, it crashes, so it should be some unhandled exception inside network event loop (based on grpcsharp_completion_queue_next method in the stack). 
I have tried attaching VS to it but it doesn't intercept any exception and Unity just closes.

The hang was annoying as well so huge thanks for that fix.
